Add config for getGPUTier method in cornerstone.init

iamiota opened the pull request:
<!-- Do Not Delete This! pr_template -->
<!-- Please read our Rules of Conduct: https://github.com/OHIF/Viewers/blob/master/CODE_OF_CONDUCT.md -->
<!-- :hand: Thank you for starting this amazing contribution! -->

<!--
⚠️⚠️ Please make sure the checklist section below is complete before submitting your PR.
To complete the checklist, add an 'x' to each item: [] -> [x]
(PRs that do not have all the checkboxes marked will not be approved)
-->

### Context
When cornerstone3D is being initialized, the `getGPUTier` method from `detect-gpu` will generate a network request, which will not be implemented within a LAN and at the same time will block the initialization of cornerstone3D.

Move the parameter of `getGPUTier` to the parameter of `cornerstone3D.init`, users can configure the path of device information by themselves. For example, they can download the device information files to a specific path on the hard disk in advance, so that even in a LAN, files can still be requested.

Perhaps the name `getGPUTierConfig` is not accurate enough, do you have any suggestions?

<!--
Provide a clear explanation of the reasoning behind this change, such as:
- A link to the issue being addressed, using the format "Fixes #ISSUE_NUMBER"
- An image showing the issue or problem being addressed (if not already in the issue)
- Error logs or callStacks to help with the understanding of the problem (if not already in the issue)
-->

### Changes & Results
```typescript
import * as cornerstone from '@cornerstonejs/core'
cornerstone.init({
  getGPUTierConfig: {
    benchmarksURL: '/resources/devices',
  },
})
```
#### Before
![WX20230526-145648@2x](https://github.com/cornerstonejs/cornerstone3D-beta/assets/7531576/da503066-5d5c-41c5-8f7d-ed0a7f06f3bc)

#### After
![WX20230526-145712@2x](https://github.com/cornerstonejs/cornerstone3D-beta/assets/7531576/cf6bc247-adbb-4286-b739-ecd76c9e2371)


<!--
List all the changes that have been done, such as:
- Add new components
- Remove old components
- Update dependencies

What are the effects of this change?
- Before vs After
- Screenshots / GIFs / Videos
-->

### Testing

<!--
Describe how we can test your changes.
- open a URL
- visit a page
- click on a button
- etc.
-->

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [] My code has been well-documented (function documentation, inline comments,
  etc.)

- [] I have run the `yarn build:update-api` to update the API documentation, and have
  committed the changes to this PR. (Read more here https://www.cornerstonejs.org/docs/contribute/update-api)


#### Public Documentation Updates

<!-- https://cornerstonejs.org/ -->

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] "OS: <!--[e.g. Windows 10, macOS 10.15.4]"--> macOS 12.6.2
- [] "Node version: <!--[e.g. 16.14.0]"--> 16.13.0
- [] "Browser: Chrome 113.0.5672.126
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]"-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

sedghi commented:
Hey, I really like the idea. But I don't see the implementation would work. I think this is what you are trying to do

the init function would be used like this 

`cornerstone3D.init({getGPUConfig: ()=>{}})` 

then inside our `init` you should check if the function is provided and use that instead of the network request and bypass the detect-gpu. 

Let me know what you think. 
Also we should update the docs to reflect this change after

iamiota commented:
I'm sorry, I didn't express myself clearly.
I don't want to generate a network request to replace `getGPUTier`.

First, I looked at the `detect-gpu` documentation and found out that the getGPUTier method allows an object parameter to be passed in, including a property called [benchmarksURL](https://github.com/pmndrs/detect-gpu#:~:text=benchmarksURL?:%20string;), which overrides the default device information request address (the default address is the request address shown in the first screenshot). 

So what I want to do is pass the benchmarksURL property to getGPUTier through the cornerstone init method. The code I submitted is effective, and the screenshot above is the result of my local verification.

I don't know if it's elegant to pass like this in the `cornerstone.init`, please forgive my poor English.
Thanks!

sbsurf commented on common/reviews/api/core.api.md:
```diff
@@ -210,6 +211,7 @@ enum ContourType {
 // @public (undocumented)
 type Cornerstone3DConfig = {
     detectGPU: any;
```
Wouldn't passing in an instance of `getGPUTier` to `detectGPU` with the `benchmarksURL` option set work?

sedghi commented:
@iamiota I see ok makes sense. Looking back at this `detectGPU` does not makes sense, it should be renamed to `gpuTier` since it holds the tier, and you can add the `detectGPUConfig` and add your benchmark url

iamiota commented:
After careful consideration, I have resubmitted some code. Overall, the description of this PR is as follows:

1. Renamed `detectGPU` to `gpuTier` and specified its type as [`TierResult`](https://github.com/pmndrs/detect-gpu/blob/master/src/index.ts#L82)(which is the return value of `getGPUTier`), it is `optional`. Users could obtain device information in their own way and pass it as a parameter to `cornerstone3D.init`. When this value exists, `cornerstone3D.init` method will use this parameter instead of making a network request to retrieve device information.

2. Added the `detectGPUConfig` parameter, which is of type [`GetGPUTier`](https://github.com/pmndrs/detect-gpu/blob/master/src/index.ts#L20) (the parameter for `getGPUTier`). This allows users to configure it based on their network conditions.

Thanks! @sedghi  @sbsurf

iamiota commented on packages/core/src/types/Cornerstone3DConfig.ts:
```diff
@@ -1,5 +1,8 @@
+
 type Cornerstone3DConfig = {
-  detectGPU: any;
+  gpuTier?: TierResult;
```
Of course, it has already been added.

sedghi commented:
You should run `yarn build:update-api` and include(commit) the results. I have done that for you this time, just FYI for next time
